### ISSUE TITLE
[Problem and a Question] Desktop shortcut is not created

### ISSUE THREAD

**Issue by Plage1984** [SteamAchievementNotifier/SteamAchievementNotifier#4]
Hi. I just wanted to note, unfortunately, there is no desktop shortcut created
Windows 10 or/and Windows 11 (both latest version), both the same.
Is there a log somewhere so I can send you possible error messages?

And the question is: The script from you, runs only as long as one is set to "Online" with the friends list on Steam. Does this also work in "Invisible" mode?

Powershell displays on Invisible: 
"
Steam Achievement Notifier:
No Steam game detected.
"
On "Online" it works then.

And thanks for your script. I've been waiting for a better achievement system on Steam for a long time.

**Comment by SteamAchievementNotifier** [comment]
Hello PlageOne,

The desktop shortcut isn't created by default, but you can create one by clicking the Settings Cog icon (⚙) and selecting "Create Desktop Shortcut". If it still doesn't create one after selecting this option, you can manually create a desktop shortcut by doing the following:

1. Right-click in an empty space on your desktop.
2. Go to New > Shortcut.
3. In "_Type the location of the item_", type: `%localappdata\SteamAchievementNotifier\SANGUI.exe` (or click "_Browse..._" and go to _This PC_ > _OSDisk (C:)_ > _Users_ > _<yourusername>_ > _AppData_ > _Local_ > _SteamAchievementNotifier_ > _SANGUI.exe_).
4. In "_Type a name for this shortcut_", name it "_Steam Achievement Notifier_" and click "_Finish_".
5. Right-click the shortcut and select "_Properties_". Click "_Change Icon..._". In the new window, click "_Browse..._". Go back to `%localappdata%\SteamAchievementNotifier` and select "_steamcheev.ico_" - this will set the correct shortcut icon.

Regarding Windows 11, the app has not been tested in a Windows 11 environment, so please be aware that I can't actively resolve any issues you may encounter outside of Windows 10 until it officially releases to the public.

And the app requires two things to function (as it checks publicly available API data to provide the notifications):
- Your Steam profile must be set to "_Public_" to read stats from the Steam Web API (this can be changed in Steam via "_Profile_" > "_Edit Profile_" > "_Privacy Settings_" > "_My Profile_"/"_Game Details_").
- Your status must be "_Online_". If you are set to "_Invisible_", your stats are also invisible. Unless Steam changes the way it's API can be accessed, there's not much I can do about these from a development perspective.

I hope this clears up your issues - please let me know if you need any further help.

Thanks!

**Comment by PlageOne** [comment]
Thanks for the explanation and answering the question.
I would like to note that nothing happens when you click on "Create Desktop Shortcut" in the settings. On Windows 10 and 11.
The way to the AppData folders of course works as usual and reliable.

Is it normal that the script "restarts" about every 15-30 minutes and then outputs this variable "Now tracking achievements for ${gamename}"? 
And that just in a loop as I said every 15-30 minutes.
Ok there I admit, this is on Windows 11, I'll have to test this again on Windows 10 later. It could be that Windows 11 has a timer in there and therefore something shoots down completely.

On Windows11:
If I run your script without game, Plopt for a fraction of a second every 15-30 minutes, a I think new Powershell window opens and then closes again immediately and that will then probably be the "problem" or whatever I should call it. 

It's very annoying but I can get over it. As long as everything else works, even under Windows 11 ;)

**Comment by SteamAchievementNotifier** [comment]
Hi again PlageOne,

That is strange behaviour that I haven't come across before. Did you download the application via the [Releases](https://github.com/SteamAchievementNotifier/SteamAchievementNotifier/releases) page, or via the "Code" > "Download ZIP" option on the [Readme](https://github.com/SteamAchievementNotifier/SteamAchievementNotifier) page? It shouldn't make a difference, but the Releases page is the intended way to download the application.

In terms of the desktop shortcut issue, I can double check what's going on if you can do the following:

- Download [this file](https://github.com/SteamAchievementNotifier/SteamAchievementNotifier/files/6980946/SANGUI-DEBUG.zip).
- Right-click > Properties > Check "Unblock" checkbox > Apply.
- Unzip the folder.
- Right-click and select "Run with Powershell".
- In the Steam Achievement Notifier GUI that appears, click the Settings Cog > "Create Desktop Shortcut", and tell me/screenshot what error you get in the Powershell console (this console window is hidden in the actual application so you wouldn't be able to see it normally).

The script should only be restarting outside of a game, so it's unusual for the "Now tracking" notification to appear multiple times. This only usually happens if your Steam client goes offline and re-connects, otherwise it will only show the notification when starting the game. As you said, though, it could be an unindentified issue with how Windows 11 interacts with Powershell or the notification module (especially if the notification literally appears as "Now tracking achievements for `${gamename}`" - this variable should always be replaced with the actual name of the game). Unfortunately, it's difficult to debug without having access to it.

Thanks!

**Comment by PlageOne** [comment]
Thanks for the feedback. Then I'll try to answer your questions.

I have downloaded your application from the release page.
The debug code as an image:

![Screenshot 2021-08-13 134007](https://user-images.githubusercontent.com/62032403/129352316-16277a81-7586-4c2b-b7e1-45a5453cc700.png)


[Quote]"  As you said, though, it could be an unindentified issue with how Windows 11 interacts with Powershell or the notification module (especially if the notification literally appears as "Now tracking achievements for ${gamename}"  [/Quote]

I have tested this in the meantime and it also happens with Windows 10, so it is not a Windows 11 problem alone. 

[Quote]"this variable should always be replaced with the actual name of the game)."[/Quote]

Yes I am aware of that.
I wanted to say that it doesn't matter which game I play, after 10-30 minutes I get this message.

Just a thought from me. Can't you build a timer into the script which scans after game start if the game process is still present?
"The Problem":

I live in Germany and we have very bad internet anyway. (Our government... Hmmm. let's not talk about that...)
I have from Vodafone Germany a 1Gbit over cable and it can just be that my connection to the Steam servers and thus to the API is disturbed for a few milliseconds.This seems to be no problem for Steam and all other programs or games but maybe for your script with the query of the API.

As I said I do not know unfortunately. Possible but unfortunately everything

Greetings

**Comment by SteamAchievementNotifier** [comment]
Hey again PlageOne,

From the screenshot, it looks like your Desktop path isn't actually set to "C:\Users\<username>\Desktop", so it can't create the shortcut there. It should be a dynamic variable (using "%userprofile%\Desktop"), so it should pick up the desktop path even if it's not located in the default location, but it doesn't seem to find it in your case. And I'm not quite sure why that would be, outside of having manually changed your environmental variables or manually setting your desktop folder to be on another drive or something.

You could manually create the shortcut as discussed before, or to troubleshoot the shortcut issue, you could do the following:
- Open the Start Menu, type `Powershell` and open.
- Type or paste `$env:userprofile` and press enter.
- Check the path.

If the resulting path isn't shown as "C:\Users\<username>" (depending on drive letter), your Desktop path won't resolve to the same place as the script intends it to. If it **does** show as "C:\Users\<username>", however, then it's going to be another issue.

> Just a thought from me. Can't you build a timer into the script which scans after game start if the game process is still present?

The way the script works at the minute is that it connects to your API stats using your credentials and retrieves the "gameid" JSON object, which is generated by the Steam Web API whenever you start a game. When a "gameid" value is found, the "Now tracking..." notification is triggered and the achievement check loop is started. This provides a simple way to keep track of your current game, but will obviously break whenever you lose connection to Steam. When reconnecting, it will see you've suddenly got a "gameid" again, and will therefore trigger another "Now tracking..." notification.

Ideally, it would be good to have some kind of fallback option using the game's process to generate the "gameid" field. However, the issue would be that as the script is web-based, it won't be able to track your achievements when you're offline due to not being able to connect to the Steam API, so having no connection would also break the rest of the script. So, unfortunately, I don't think there's a way around that issue for now without adding an extensive rework of the script itself, but ideally, I would love to have the script run locally so it might be a possibility in the future.

If you don't care about showing the "Now tracking..." notification at all and just want to track the achievements, you can manually comment out the line that triggers it by doing the following:

- Open the Start menu and type `%localappdata%\SteamAchievementNotifier`.
- Right-click "SteamAchievementNotifier.ps1" and click "Edit". (_This should open in an application called "Powershell ISE"._)
- Find line 49, which should be: `New-BurntToastNotification -UniqueIdentifier "Track" -Silent -AppLogo $steam -Text "Steam Achievement Notifier","Now tracking achievements for ${gamename}.";`.
- At the beginning of the line, type a "#". The whole line should now be: `#New-BurntToastNotification -UniqueIdentifier "Track" -Silent -AppLogo $steam -Text "Steam Achievement Notifier","Now tracking achievements for ${gamename}.";`.
- Click "File" > "Save".
- Start Steam Achievement Notifier as normal.

This will no longer show the "Now tracking..." notification and will hopefully work around the issue.

Thanks!

**Comment by PlageOne** [comment]
Hi hi,

Here is the result:

PS C:\Users\plage> $env:userprofile
C:\Users\plage
PS C:\Users\plage>

But it just occurred to me and I could have thought of it myself.... (Facepalm to me)
When installing Windows you are asked if you want to sync selected data with OneDrive and if you use OneDrive you do that. From this point the default path of "Documents", "Pictures" and also "Desktop" is changed to:

"C:\Users\<username>\OneDrive\Desktop".

I'm sorry, I honestly did not think of it. Will then probably be because of it or?


to the "Now tracking..." message: 

Yes that's understandable, then I will comment that out now.
Thanks for your explanation and especially for your help :)

**Comment by SteamAchievementNotifier** [comment]
Hey again PlageOne,

Well spotted - looks like that's going to be your issue!

The GUI is looking for your Desktop in "C:\Users\username" but if your Desktop actually points to "C:\Users\username\OneDrive\Desktop", it won't be able to find it. (_You can check the current location of your Library folders (Desktop/Documents/Music etc.) by right-clicking them in Explorer > Properties > Location tab. You can change the folder location from here, but I wouldn't recommend it if you've already got files in the OneDrive location._)

The simplest way to fix this would be to manually edit the script again:
- Double-check your current Desktop folder location by opening Windows Explorer, right-clicking "Desktop" in the left-side Navigation Pane, selecting "Properties" and clicking the "Location" tab. Take note of this address.
- Open the Start menu and type/paste: `%localappdata%\SteamAchievementNotifier`.
- When the folder opens, right-click "_systraygui.ps1_" > "Edit".
- When Powershell ISE opens, find: 
> **Line 324**
> `if (!(Test-Path "$env:userprofile\Desktop\Steam Achievement Notifier.lnk" -PathType Leaf)) {`
> **Line 327**
> `$shortcut = $shell.CreateShortcut("$env:userprofile\Desktop\Steam Achievement Notifier.lnk")`

- Replace the path "**$env:userprofile\Desktop\Steam Achievement Notifier.lnk**" with your actual Desktop path - presumably, in your case: "**C:\Users\Plage\OneDrive\Desktop\Steam Achievement Notifier.lnk**". It should now look like this (assuming the path is correct):
> **Line 324**
> `if (!(Test-Path "C:\Users\Plage\OneDrive\Desktop\Steam Achievement Notifier.lnk" -PathType Leaf)) {`
> **Line 327**
> `$shortcut = $shell.CreateShortcut("C:\Users\Plage\OneDrive\Desktop\Steam Achievement Notifier.lnk")`
- Once done, click "File" > "Save". Then try opening Steam Achievement Notifier again and creating a Desktop shortcut.

I'll add this bug to the fix list and hopefully eliminate the issue in the next release.

Let me know how it goes - always happy to help!

Thanks!

**Comment by PlageOne** [comment]
Hello again :),

Shortcut:

Changing it to "~\OneDrive\Desktop\" works and creating the shortcut to the desktop now works with it.

"Now tracking achievements for":

I opened the script directly with Windows Powershell ISE and started it. I had very unstable internet yesterday, but still I had no more problems despite reconnects and packet loss. Your script worked very well, but only when I start it via ISE.
If I start it normally, exactly what I have already mentioned happens.

Greetings

**Comment by SteamAchievementNotifier** [comment]
Hey again PlageOne,

I'm glad one issue has been resolved now! I'm currently working on Version 1.6, which includes a check for the actual path of your Desktop, instead of assuming it is always going to be "C:\Users\_username_\Desktop". This should be released very soon, so be sure to check the Releases page.

In regards to the "Now tracking..." notification issue, if you find that it still happens in 1.6, I'll look into the issue further. It's weird that it would behave differently in ISE, but maybe it's something I've overlooked.

I'll keep this issue open until 1.6 releases so we can see whether it still happens.

Thanks!

**Comment by PlageOne** [comment]
Hi,
I am back and have tested version 1.6. 
In desktop mode, so if no game is started, no message comes and that's good.

When playing, however, it still happens from time to time that I sporadically get the said message. But I think that is also mainly due to my Internet, since you will probably can not do anything against I think.
